std/arith.asm

```diff
@@ -150,44 +156,125 @@ machine Arith(CLK32_31, operation_id){
     sum(16, |i| x1[i] * CLK32[i]) + sum(16, |i| y1[i] * CLK32[16 + i]) in BYTE2;
     sum(16, |i| x2[i] * CLK32[i]) + sum(16, |i| y2[i] * CLK32[16 + i]) in BYTE2;
     sum(16, |i| x3[i] * CLK32[i]) + sum(16, |i| y3[i] * CLK32[16 + i]) in BYTE2;
+    // Note that for q0-q2, we only range-constrain the first 15 limbs here
+    sum(16, |i| s[i] * CLK32[i]) + sum(15, |i| q0[i] * CLK32[16 + i]) in BYTE2;
+    sum(15, |i| q1[i] * CLK32[i]) + sum(15, |i| q2[i] * CLK32[16 + i]) in BYTE2;
+
+    // The most significant limbs of q0-q2 are constrained to be 32 bits
+    // In Polygon's version they are 19 bits, but that requires increasing the minimum degree
+    // to fit the lookup.
+    // Instead, we decompose the most significant limb into two 16-Bit limbs.
+    // Having a larger range-constraint is fine, because we're only multiplying it with 16-bit
+    // limbs of the prime, so the result is within 48 bits, still far from overflowing the
+    // Goldilocks field.
+    pol witness q0_15_high, q0_15_low, q1_15_high, q1_15_low, q2_15_high, q2_15_low;
+    q0_15_high * CLK32[0] + q0_15_low * CLK32[1] + q1_15_high * CLK32[2] + q1_15_low * CLK32[3] + q2_15_high * CLK32[4] + q2_15_low * CLK32[5] in BYTE2;
+
+    unchanged_until(q0_15_high, CLK32[31]);
```

Comment:
Maybe it makes sense to have a special function `|x| unchanged_until(x, CLK32[31])`?

Reply:
Done.